### PR TITLE
Add a generic cell list implementation

### PR DESCRIPTION
**Pull request by rc83**
The implementation is based on templates that allows both versatility and compiler optimizations.

* various grid types: fixed and periodic
* various containers: dense and sparse

# Description

I like the double inheritance of the cell list from both the grid and the container which allows a great flexibility and also simplifies unit testing. Abstract classes, however, are needed as an interface.

I am not so sure about using decorators and mixin-like extensions to create various flavours of cell lists. Currently it is broken down to very elementary pieces – with an idea to simplify unit testing, but it has turned out that it is not the case. Perhaps more monolithic approach can make the code more comprehensible.

Discussion welcome.

## Checklist

- [ ] `make test` passes with no errors
- [ ] the source code is well documented
- [ ] new functionality includes unittests
- [ ] the user-manual has been updated to reflect the changes
- [ ] I have installed the provided commit hook to auto-format commits (requires `clang-format`):

  ``` bash
  ./scripts/git-pre-commit-format install
  ```

- [ ] code naming scheme follows the convention:

  Style        | Elements
  ------------ | ---------------------
  `PascalCase` | classes, namespaces
  `camelCase`  | functions
  `snake_case` | variables

### CODE REVIEW

File: src/celllist.h [hunk 1]
```diff
@@ -1,161 +1,164 @@
...
- *   is obtained with `neighbors()`
- * - index can be moved from one grid point to another with `move()`
- * - the list of particle index in each grid point is stored in a
- *   `std::set<int>` container.
+ * @tparam VDimension  number of grid dimensions, e.g., 3 for the real 3D world
+ * @tparam TCellIndex  type of a single coordinate in cell space; also an absolute cell index
+ * @tparam TSpaceAxis  type of a single coordinate in physical space
+ */
+template <unsigned int VDimension, typename TCellIndex = int, typename TSpaceAxis = double> struct GridType {
+    constexpr static unsigned int Dimension = VDimension; //!< number of dimmensions, e.g., 3 for the real 3D world
```
Comment (mlund) [inline src/celllist.h:34]: Why not use template value directly?

File: src/celllist.h [hunk 2]
```diff
@@ -1,161 +1,164 @@
...
+     */
+    GridOffsets3D(CellIndex distance);
+
+  private:
+    void initNeighbors();
+    void initForwardNeighbors();
+};
+} // namespace Grid
+
+namespace Container {
```
Comment (mlund) [inline src/celllist.h:67]: add namespace documentation

File: src/celllist.h [hunk 3]
```diff
@@ -1,161 +1,164 @@
...
-    typedef Eigen::Vector3d Point;
-    Point halfbox;
-    double cellsize = 0;                                           // cell side length (angstrom)
-    std::vector<std::vector<std::vector<std::set<Tindex>>>> cells; // dense storage
-
-    std::vector<std::set<Tindex>> cells_dense;    // dense storage (to replace `cells`)
-    std::map<int, std::set<Tindex>> cells_sparse; // sparse storage
+template <typename TMember, typename TIndex> struct AbstractContainer {
+    using ContainerType = AbstractContainer<TMember, TIndex>;
+    typedef TIndex Index;  //!< the cell index (key) type
```
Comment (mlund) [inline src/celllist.h:87]: Why introduce extra names?

File: src/celllist.h [hunk 4]
```diff
@@ -1,161 +1,164 @@
...
+};
+} // namespace Container
 
-    inline std::set<Tindex> &row_major(const CellPoint &c) {
-        return cells_dense[c[0] * KLM[0] * KLM[2] + c[1] * KLM[2] + c[2]]; // to replace operator[] below
-    }                                                                      // row-major ordering of dense storage
+/**
+ * Type manipulation helpers.
+ */
+template <typename T> using ContainerOf = typename T::Container;
```
Comment (mlund) [inline src/celllist.h:107]: What does "Of" mean?

File: src/celllist.h [hunk 5]
```diff
@@ -1,161 +1,164 @@
...
-        assert(cnt == 27);
-    } //!< Index from all 26+1 neighboring+own cells (complexity: N neighbors)
+/**
+ * @brief An interface of a immutable cell list. No methods for manipulating (adding, removing) members are provided.
+ * @tparam TContainerType  a container type to obtain Member and Members subtypes
+ * @tparam TGridType  a grid type to obtain index and coordinates subtypes
+ */
+template <class TContainerType, class TGridType> struct AbstractImmutableCellList {
+    using AbstractCellList = AbstractImmutableCellList<TContainerType, TGridType>; //! self
+    using ContainerType = TContainerType;                                          //!< abstract container types
```
Comment (mlund) [inline src/celllist.h:125]: why?

File: src/celllist.h [hunk 6]
```diff
@@ -1,161 +1,164 @@
...
-#endif
+/**
+ * @brief An interface to a cell list that can return cell's members in sorted order.
+ *
+ * Interface is used to efficiently compute differences between cell lists.
+ *
+ * @tparam TContainerType  a container type to obtain Member and Members subtypes
+ * @tparam TGridType  a grid type to obtain index and coordinates subtypes
+ */
+template <class TContainerType, class TGridType>
```
Comment (mlund) [inline src/celllist.h:152]: What is the purpose of "T" prefix?

File: src/celllistimpl.cpp [hunk 7]
```diff
@@ -0,0 +1,263 @@
+#include <doctest/doctest.h>
+#include <type_traits>
+#include "celllistimpl.h"
+namespace Faunus {
+namespace CellList {
+namespace Grid {
+
```
Comment (mlund) [inline src/celllistimpl.cpp:7]: Since c++17 the is allowed, `namespace Faunus::CellList::Grid {`

File: src/celllistimpl.cpp [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+    const int neighbors_count = std::pow(1 + 2 * distance, 3) - 1;
+    neighbors.clear();
+    neighbors.reserve(neighbors_count);
+    for (auto i = -distance; i <= distance; ++i) {
+        for (auto j = -distance; j <= distance; ++j) {
+            for (auto k = -distance; k <= distance; ++k) {
+                if (i == 0 && j == 0 && k == 0) {
+                    continue;
+                }
+                neighbors.push_back({i, j, k});
```
Comment (mlund) [inline src/celllistimpl.cpp:23]: `emplace_back()`

File: src/celllist.h [hunk 9]
```diff
@@ -1,161 +1,164 @@
...
+ *
+ * @author Richard Chudoba
+ * @date 2021-02-01
+ */
 
 namespace Faunus {
 
+namespace CellList {
 
+namespace Grid {
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/celllist.h:16]: Also `namespace Faunus::CellList::Grid {}` is allowed since c++17.

File: src/celllistimpl.h [hunk 10]
```diff
@@ -0,0 +1,911 @@
...
+     */
+    bool isCell(const CellCoord&) const override { return true; }
+
+    /**
+     * @brief Verifies if spatial coordinates are in the box.
+     *
+     * Any coordinates are valid under full PBC.
+     *
+     * @param point
+     * @return  always true
```
Comment (mlund) [inline src/celllistimpl.h:334]: why?

Comment (mlund) [reply src/celllistimpl.h:334]: error if really outside?

**Review by mlund — APPROVED** [review]

File: src/celllistimpl.cpp [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+            CHECK((grid.coordinatesAt({3., 3., 3.}) == CellCoord{1, 2, 1}).all());
+        }
+    }
+}
+
+TEST_CASE("Grid3DPeriodic") {
+    using TestGridBase = Grid3DPeriodic;
+    using CellCoord = TestGridBase::CellCoord;
+    TestGridBase grid({20., 30., 40.}, 1.5);
+    CellCoord end{(200 / 15), (300 / 15), (400 / 15)};
```
Comment (mlund) [inline src/celllistimpl.cpp:105]: integer division intended?

**Review by mlund — APPROVED** [review]

File: src/celllistimpl.h (hunk 10 above)
Comment (rc83) [reply src/celllistimpl.h:334]: To check if the given coordinates, computed for example as cell + offset, can be translated to a cell index. That is always possible under periodic boundary conditions, but not under fixed boundary conditions, e.g., for the cells at edges. That is the sole purpose of the grid to hide boundary condition details. Documentation extended accordingly.

File: src/celllistimpl.cpp (hunk 11 above)
Comment (rc83) [reply src/celllistimpl.cpp:105]: Yes. Cell coordinates are an ordered tuple of integers (`CellCoord` is `Vector3i`).

File: src/celllist.h (hunk 1 above)
Comment (mlund) [reply src/celllist.h:34]: `Dimension` --> `dimension`

File: src/celllistimpl.h [hunk 12]
```diff
@@ -0,0 +1,911 @@
...
+ * the expense of a bigger memory footprint, expecialy in sparse systems.
+ *
+ * @tparam TMember  member (value)
+ * @tparam TIndex  index (key); corresponds do the CellIndex of GridType
+ */
+template <typename TMember, typename TIndex> class DenseContainer : virtual public AbstractContainer<TMember, TIndex> {
+  public:
+    using ContainerType = AbstractContainer<TMember, TIndex>;
+    using typename ContainerType::Index;
+    using typename ContainerType::Members;
```
Comment (mlund) [inline src/celllistimpl.h:405]: add comment for Mikael

File: src/celllistimpl.h [hunk 13]
```diff
@@ -0,0 +1,911 @@
...
+
+    /**
+     * @brief Return all cell indices that may contain members.
+     *
+     * It is not guaranteed that all returned cells are non-empty though. However all non-empty cells are returned.
+     *
+     * @return  vector of cell indices
+     */
+    std::vector<Index> indices() const override {
+        std::vector<Index> indices;
```
Comment (mlund) [inline src/celllistimpl.h:427]: Would reserve make sense?

File: src/celllistimpl.h [hunk 14]
```diff
@@ -0,0 +1,911 @@
...
+ * @tparam TMember  member (value)
+ * @tparam TIndex  index (key); corresponds do CellIndex of GridBase
+ */
+template <typename TMember, typename TIndex> class SparseContainer : virtual public AbstractContainer<TMember, TIndex> {
+  public:
+    using ContainerType = AbstractContainer<TMember, TIndex>;
+    using typename ContainerType::Index;
+    using typename ContainerType::Members;
+
+    const Members& get(Index index) const override {
```
Comment (mlund) [inline src/celllistimpl.h:461]: add `@throw` documentation

File: src/celllistimpl.h [hunk 15]
```diff
@@ -0,0 +1,911 @@
...
+    /**
+     * @brief Return all cell indicis that may contain members.
+     *
+     * It is not guaranteed that all selected cells are non-empty though.
+     *
+     * @return  vector of cell indices
+     */
+    std::vector<Index> indices() const override {
+        std::vector<Index> indices;
+        transform(container.begin(), container.end(), back_inserter(indices),
```
Comment (mlund) [inline src/celllistimpl.h:489]: add `std::`

File: src/celllistimpl.h [hunk 16]
```diff
@@ -0,0 +1,911 @@
...
+    using ContainerType = ContainerTypeOf<TContainer>;
+    using Member = MemberOf<ContainerType>;
+    using Index = IndexOf<ContainerType>;
+
+    void add(const Member& member, const Index index_new) {
+        assert(index_new < this->indexEnd());
+        this->get(index_new).push_back(member);
+    }
+
+    void remove(const Member& member, const Index index_old) {
```
Comment (mlund) [inline src/celllistimpl.h:522]: better with `erase` as in STL?

File: src/celllistimpl.h [hunk 17]
```diff
@@ -0,0 +1,911 @@
...
+        } else {
+            return getEmpty();
+        }
+    }
+
+    /**
+     * @return  range of containg coordinates of all possibly non-empty cells
+     */
+    std::vector<CellCoord> getCells() const {
+        auto vec = this->indices();
```
Comment (mlund) [inline src/celllistimpl.h:589]: vec -> indices?

File: src/celllistimpl.h [hunk 18]
```diff
@@ -0,0 +1,911 @@
...
+    /**
+     *  @brief Construct an empty cell list knowing a box dimension and minimal cell dimension.
+     *  @param box  spatial dimension of the box
+     *  @param cell_dimension  minimal length of the cell's edge
+     */
+    CellListBase(const typename TGrid::Point& box, const typename TGrid::SpaceAxis cell_dimension)
+        : Grid(box, cell_dimension), Container(Grid::size()) {}
+
+  protected:
+    //    const auto& getMembers(const CellIndex cell_index) const { return this->get(cell_index); }
```
Comment (mlund) [inline src/celllistimpl.h:623]: remove?

File: src/celllistimpl.h [hunk 19]
```diff
@@ -0,0 +1,911 @@
...
+     * @param cell_index
+     * @return  sorted members
+     */
+    const Members& getSorted(const CellIndex cell_index) {
+        auto& members = this->get(cell_index);
+        if (!is_sorted[cell_index]) {
+            std::sort(members.begin(), members.end());
+            is_sorted[cell_index] = true;
+        }
+        assert(std::is_sorted(members.begin(), members.end()));
```
Comment (mlund) [inline src/celllistimpl.h:728]: assertion not really needed

File: src/celllistimpl.h [hunk 20]
```diff
@@ -0,0 +1,911 @@
...
+        return members;
+    }
+
+  private:
+    // TODO perhaps a map for sparse containers
```
Comment (mlund) [inline src/celllistimpl.h:733]: I think vector is fine; the STL implementation is supposedly using optimized bit storage for std::vector<bool>

File: src/celllistimpl.h [hunk 21]
```diff
@@ -0,0 +1,911 @@
...
+        member2cell.erase(member);
+    }
+
+    void updateMember(const Member& member, const CellCoord& new_cell_coordinates) {
+        this->update(member, this->index(new_cell_coordinates));
+    }
+
+    /**
+     * @brief Imports members from other list without computing cell coordinates from member positions.
+     * @tparam T
```
Comment (mlund) [inline src/celllistimpl.h:768]: fill in or remove

File: src/celllistimpl.h [hunk 22]
```diff
@@ -0,0 +1,911 @@
...
+        for (const auto& m : members) {
+            add(m, source.member2cell[m]);
+        }
+    }
+
+    using TBase::TBase;
+
+  protected:
+    void add(const Member& member, const CellIndex& new_cell_index) {
+        assert(member2cell.count(member) == 0); // FIXME C++20 contains
```
Comment (mlund) [inline src/celllistimpl.h:782]: test for c++20 using macro?

File: src/celllistimpl.h [hunk 23]
```diff
@@ -0,0 +1,911 @@
...
+    using CellListReverseMap<TBase>::CellListReverseMap;
+};
+
+/**
+ * @brief Wrapper of two cell list to efficiently provide difference of members, i.e., members presented in the minuend
+ * and not in the subtrahend.
+ *
+ * @todo Allow cache invalidation when underlying cell lists change
+ *
+ * @tparam TCellListMinuend
```
Comment (mlund) [inline src/celllistimpl.h:832]: describe

File: src/celllistimpl.h [hunk 24]
```diff
@@ -0,0 +1,911 @@
...
+                auto minuend_cell = minuend->getSortedMembers(cell_coordinates);
+                std::tie(difference_cell_it, std::ignore) = difference_cache.emplace(cell_index, Members{});
+                std::set_difference(minuend_cell.begin(), minuend_cell.end(), subtrahend_cell.begin(),
+                                    subtrahend_cell.end(), std::back_inserter(difference_cell_it->second));
+            }
+            return difference_cell_it->second;
+        }
+    }
+
+    const Members& getMembers(const CellCoord& cell_coordinates) override {
```
Comment (mlund) [inline src/celllistimpl.h:869]: const function?

File: src/celllistimpl.h [hunk 25]
```diff
@@ -0,0 +1,911 @@
...
+
+    auto& getGrid() { return minuend->getGrid(); }
+
+    std::vector<CellCoord> getCells() const override { return minuend->getCells(); }
+
+    CellListDifference(std::shared_ptr<TCellListMinuend> minuend, std::shared_ptr<TCellListSubtrahend> subtrahend)
+        : minuend(minuend), subtrahend(subtrahend) {}
+
+  private:
+    std::shared_ptr<CellListMinuend> minuend;
```
Comment (mlund) [inline src/celllistimpl.h:905]: these terms were new to me

File: src/celllistimpl.h [hunk 26]
```diff
@@ -0,0 +1,911 @@
...
+    void add(const Member& member, const CellIndex& new_cell_index) {
+        TBase::add(member, new_cell_index);
+        if (is_sorted[new_cell_index]) {
+            is_sorted[new_cell_index] = false;
+        }
+    }
+
+    void remove(const Member& member, const CellIndex& old_cell_index) {
+        TBase::remove(member, old_cell_index);
+        if (is_sorted[old_cell_index]) {
```
Comment (mlund) [inline src/celllistimpl.h:712]: why bother testing instead of setting it to `false`

File: src/celllistimpl.h [hunk 27]
```diff
@@ -0,0 +1,911 @@
...
+    SortableCellList(const typename GridOf<TBase>::Point& box, const typename GridOf<TBase>::SpaceAxis cell_dimension)
+        : TBase(box, cell_dimension) {
+        is_sorted.resize(this->size(), false);
+    }
+
+  protected:
+    void add(const Member& member, const CellIndex& new_cell_index) {
+        TBase::add(member, new_cell_index);
+        if (is_sorted[new_cell_index]) {
+            is_sorted[new_cell_index] = false;
```
Comment (mlund) [inline src/celllistimpl.h:706]: why test?

File: src/celllistimpl.h [hunk 28]
```diff
@@ -0,0 +1,911 @@
...
+  private:
+    CellCoord neighbours_cell_offset_min; //!< utmost negative offset of neighbours to prevent circular match in PBC
+    CellCoord neighbours_cell_offset_max; //!< utmost positive offset of neighbours to prevent circular match in PBC
+
+    /**
+     * @brief Setup immutable parameters based on the box dimensions. To be called in constructors only.
+     * @param new_box
+     */
+    void updateCellGrid() {
+        const auto cell_list_middle = (this->getCellListEnd() - 1).template cast<double>() / 2.;
```
Comment (mlund) [inline src/celllistimpl.h:363]: don't know if this optimized out, but I'd prefer multiplication with 0.5

File: src/celllistimpl.h [hunk 29]
```diff
@@ -0,0 +1,911 @@
...
+/**
+ * @brief Interface of a grid
+ * @tparam TGridType  a cell grid types declaration
+ */
+template <typename TGridType> struct AbstractGrid : private TGridType {
+    using GridType = TGridType;
+    using typename GridType::CellCoord;
+    using typename GridType::CellIndex;
+    using typename GridType::Point;
+    using typename GridType::SpaceAxis;
```
Comment (mlund) [inline src/celllistimpl.h:69]: I understand the logic, but got confused a few times about the "SpaceAxis". Can we find a name that makes one think in the direction of a floating point type? Also, it's difficult to imagine other types than float or double, so perhaps we can simply make a typedef? One advantage is that the IDE will be able to make better code analysis.

File: src/celllistimpl.cpp [hunk 30]
```diff
@@ -0,0 +1,263 @@
...
+namespace CellList {
+namespace Grid {
+
+GridOffsets3D::GridOffsets3D(CellIndex distance) : distance(distance) {
+    initNeighbors();
+    initForwardNeighbors();
+}
+
+void GridOffsets3D::initNeighbors() {
+    const int neighbors_count = std::pow(1 + 2 * distance, 3) - 1;
```
Comment (mlund) [inline src/celllistimpl.cpp:14]: int --> size_t

File: src/celllistimpl.cpp [hunk 31]
```diff
@@ -0,0 +1,263 @@
...
+    std::copy_if(neighbors.begin(), neighbors.end(), std::back_inserter(forward_neighbors), [](auto& offset) {
+        return (offset > CellCoord{0, 0, 0}).all();
+    });
+    assert(forward_neighbors.size() == forward_neighbors_count);
+}
+
+/**
+ * Tests
+ */
+TEST_CASE("Grid3DFixed") {
```
Comment (mlund) [inline src/celllistimpl.cpp:43]: Thorough unittesting, nice!

File: src/celllistimpl.h [hunk 32]
```diff
@@ -0,0 +1,911 @@
...
+  public:
+    using ContainerType = AbstractContainer<TMember, TIndex>;
+    using typename ContainerType::Index;
+    using typename ContainerType::Members;
+
+    const Members& get(Index index) const override {
+        assert(index >= 0 && index < indexEnd());
+        try {
+            return container.at(index);
+        } catch (std::out_of_range& e) { return empty_set; }
```
Comment (mlund) [inline src/celllistimpl.h:465]: Another alternative is to use `std::optional`

File: src/celllistimpl.h [hunk 33]
```diff
@@ -0,0 +1,911 @@
...
+    void remove(const Member& member, const Index index_old) {
+        assert(index_old < this->indexEnd());
+        auto& members = this->get(index_old);
+        // std::erase_if available in C++20
+        auto remove_it = std::remove(members.begin(), members.end(), member);
+        members.erase(remove_it, members.end());
+        // end of std::erase_if
+    }
+
+    void update(const Member& member, const Index index_old, const Index index_new) {
```
Comment (mlund) [inline src/celllistimpl.h:531]: Difficult to see purpose from signature; add documentation

File: src/celllistimpl.h [hunk 34]
```diff
@@ -0,0 +1,911 @@
...
+ * @tparam TIndex
+ * @tparam TContainer
+ */
+template <class TContainer> class Container : public TContainer {
+  public:
+    using ContainerType = ContainerTypeOf<TContainer>;
+    using Member = MemberOf<ContainerType>;
+    using Index = IndexOf<ContainerType>;
+
+    void add(const Member& member, const Index index_new) {
```
Comment (mlund) [inline src/celllistimpl.h:517]: rename to "insert"

File: src/celllistimpl.h [hunk 35]
```diff
@@ -0,0 +1,911 @@
...
+/**
+ * @brief A mixin allowing to add and remove members into and from the container, respectively.s
+ * @tparam TMember
+ * @tparam TIndex
+ * @tparam TContainer
+ */
+template <class TContainer> class Container : public TContainer {
+  public:
+    using ContainerType = ContainerTypeOf<TContainer>;
+    using Member = MemberOf<ContainerType>;
```
Comment (mlund) [inline src/celllistimpl.h:514]: add documentation

File: src/celllistimpl.h [hunk 36]
```diff
@@ -0,0 +1,911 @@
...
+    }
+
+    /**
+     * @brief Imports members from other list without computing cell coordinates from member positions.
+     * @tparam T
+     * @param source
+     * @param members
+     */
+    template <typename T> void importMembers(CellListReverseMap& source, const T& members) {
+        for (const auto& m : members) {
```
Comment (mlund) [inline src/celllistimpl.h:773]: matter of taste, but `std::for_each`?

File: src/celllistimpl.h [hunk 37]
```diff
@@ -0,0 +1,911 @@
...
+            add(m, source.member2cell[m]);
+        }
+    }
+
+    using TBase::TBase;
+
+  protected:
+    void add(const Member& member, const CellIndex& new_cell_index) {
+        assert(member2cell.count(member) == 0); // FIXME C++20 contains
+        TBase::add(member, new_cell_index);
```
Comment (mlund) [inline src/celllistimpl.h:783]: another argument for renaming add --> insert.

File: src/celllistimpl.h [hunk 38]
```diff
@@ -0,0 +1,911 @@
...
+        member2cell.insert({member, new_cell_index});
+    }
+
+    void update(const Member& member, const CellIndex& new_cell_index) {
```
Comment (mlund) [inline src/celllistimpl.h:787]: rename to move()

File: src/celllistimpl.h [hunk 39]
```diff
@@ -0,0 +1,911 @@
...
+ */
+template <class TBase> class CellListSpatial : public CellListReverseMap<TBase> {
+  public:
+    using typename CellListReverseMap<TBase>::AbstractCellList; //!< a structure with types definitions
+    using typename CellListReverseMap<TBase>::Grid;
+    using typename CellListReverseMap<TBase>::Container;
+    using typename AbstractCellList::Member;          //!< member type
+    using typename AbstractCellList::GridType::Point; //!< point
+
+    void addMemberAt(const Member& member, const Point& new_position) {
```
Comment (mlund) [inline src/celllistimpl.h:813]: insertMemberAt()

File: src/celllistimpl.h [hunk 40]
```diff
@@ -0,0 +1,911 @@
...
+        const auto new_cell_index = this->indexAt(new_position);
+        this->add(member, new_cell_index);
+    }
+
+    void updateMemberAt(const Member& member, const Point& new_position) {
```
Comment (mlund) [inline src/celllistimpl.h:818]: moveMemberToPosition()

**Pull request merged**